[Docs] Fix the table of contents and edit content (#28500)
gravitational/teleport: docs/pages/application-access/getting-started.mdx

videoBanner: 5Uwhp3IQMHY
---

> This tutorial demonstrates how to configure secure access to an application through Teleport. 
> The tutorial uses Grafana as a sample application because it's straightforward to install and run 
> in a Docker container or Kubernetes cluster with no additional configuration required.
> If you want to configure access for a different web application, you can use this tutorial as a 
> general guide for what to do. 

Reviewer: I don't think we've documented it anywhere, but we put hard line breaks in the docs around 80 characters, to make viewing/editing the source easy regardless of the editor being used:

```suggestion
This tutorial demonstrates how to configure secure access to an application
through Teleport. The tutorial uses Grafana as a sample application because
it's straightforward to install and run in a Docker container or Kubernetes
cluster with no additional configuration required. If you want to configure
access for a different web application, you can use this tutorial as a general
guide for what to do. 
```

This comment applies throughout.

Author: Yeah, I see. I generally have followed the one sentence per line rule instead. But I'm flexible.

At a high level, configuring access for applications involves the following steps:

* Verify your environment meets the prerequisites.
* Verify you can launch the application in a Docker container, Kubernetes cluster, or using another method.

> * Generate a short-lived token for the application.

Reviewer: The short-lived token is a join token for the Teleport Application Service to join the Teleport cluster.

> * Install and configure on the application host.

Reviewer: Is this sentence missing a subject? What's being installed and configured on the application host?

> * Verify your environment meets the prerequisites.
> * Verify you can launch the application in a Docker container, Kubernetes cluster, or using another method.
> * Generate a short-lived token for the application.
> * Install and configure on the application host.
> * Add a user to verify access to the application.

Reviewer: I believe our linter pings us when we use more than one markdown style for bullet points. Generally, we use `-`.

```suggestion
- Verify your environment meets the prerequisites.
- Verify you can launch the application in a Docker container, Kubernetes cluster, or using another method.
- Generate a short-lived token for the application.
- Install and configure on the application host.
- Add a user to verify access to the application.
```

If you have not yet deployed the Auth Service and Proxy Service, see one of the [getting started guides](../get-started.mdx) or try the

> Teleport Application Access [interactive learning track](https://play.instruqt.com/teleport/invite/rgvuva4gzkon).

Reviewer: While we're here, this is a good opportunity to callout the Teleport Team free trial.

> The following diagram illustrates the basic steps for connecting to Grafana using Teleport.

Reviewer: ```suggestion
The following diagram illustrates the basic steps for connecting to Grafana using Teleport:
```

videoBanner: 5Uwhp3IQMHY
---

This tutorial demonstrates how to configure secure access to an application 
through Teleport. The tutorial uses Grafana as a sample application because 
it's straightforward to install and run in a Docker container or Kubernetes cluster 
with no additional configuration required. If you want to configure access 
for a different web application, you can use this tutorial as a 
general guide for what to do. 

At a high level, configuring access for applications involves the following steps:

- Verify your environment meets the prerequisites.
- Verify you can launch the application in a Docker container, Kubernetes cluster, or 
  using another method.
- Generate a short-lived invitation token for the application to join the Teleport cluster.
- Install and configure Teleport on the application host.
- Add a user to verify access to the application.

> If you have not yet deployed the Auth Service and Proxy Service, see one of the 
> [getting started guides](../get-started.mdx), sign up for a free trial of 
> [Teleport Team](https://goteleport.com/signup/), or try the Teleport Application Access 
> [interactive learning track](https://play.instruqt.com/teleport/invite/rgvuva4gzkon).

Reviewer 2: ```suggestion
```

This repeats the Prerequisites

- (!docs/pages/includes/tctl.mdx!)
- A host where you will run the Teleport Application Service.

> - A DNS subdomain for each application.

Reviewer 3: Most people will just use a wildcard DNS here - something like `*.teleport.example.com`.

Author: Yep, there's actually an include box that mentions this. I'm not a big fan of making these common "include" snippets in these assertive little boxes, but I didn't want to rock the boat **too** much in my first PR 😄 .
![image](https://github.com/gravitational/teleport/assets/136391445/ff670e08-5d9b-4db4-a6fa-6c9900ca8dc7)

   
   <Admonition type="note" title="Application Access and DNS for Self-Hosted Clusters">

   (!docs/pages/includes/dns-app-access.mdx!)

   Teleport Cloud will automatically provide a subdomain and signed TLS certificate for your application under your tenant address.
   </Admonition>

- Docker or Kubernetes to launch Grafana.

   This tutorial illustrates how to launch Grafana in a Docker container or Kubernetes cluster.

>    Therefore, you must have either Docker or Kubernetes installed. For more information, click **Docker** or **Kubernetes cluster**.

Reviewer: I suggest removing this paragraph as unnecessary. Docker/Kubernetes is mentioned in the second sentence of the first paragraph of the page, again in the 2nd bullet point of the first list, and directly above and below. I'm not sure this adds anything that isn't already explained.

   
   <Tabs>
   <TabItem label="Docker">

>    You must have some version of Docker installed to launch Grafana in a Docker container. 

Reviewer: This line replaces "A Docker installation, which we will use to launch Grafana in a container.". I think the original is more succinct and less ambiguous (some version).

Author: Not a fan of the which clause, first person plural, or future tense. But I'm thinking of. a slightly different approach to the text here. It seemed odd to me that the Kubernetes tab mentions specific version information and the Docker tab didn't. My initial thought was to make the tabs more parallel, but yeah, that wasn't useful at all.

   If you are configuring access to a different web application using Teleport, use the appropriate commands to 
   launch that application in a Docker container.
   </TabItem>
   <TabItem label="Kubernetes cluster">

   (!docs/pages/includes/kubernetes-access/helm-k8s.mdx!)

   - Administrative rights to deploy within a cluster.

   </TabItem>
   </Tabs>

- Check that you are using appropriate permissions.
   The permissions used in examples in this tutorial are intended for demo environments.
   Before running Teleport in a production environment, check that you are following best practices to avoid security incidents.

   (!docs/pages/includes/permission-warning.mdx!)

## Step 1/5. Generate a token

A join token is required to authorize a Teleport Application Service instance to

> join the cluster. Generate a short-lived join token and save it, for example,

Reviewer: While we're in here, we should specify that the join token is being saved on the host where we will run the Teleport Application Service.

Author: I think the instructions are a little vague here. I took a stab at making them somewhat more specific. It seems like I needed to run this particular command on the instance where the Auth service runs. To run it on an application service host seems like it would require additional command-line options, like --identity and --auth-server, but I'm not sure how you create an identity file without digging in deeper, so I modified the description around this a little.

Reviewer: The `tctl` command can be run on the auth service or the local workstation after logging in with `tsh`. We should specify the latter. The point I'm making is that the file we save the token string to should be on the application service host, so that when we start that service it can read the token.

- (!docs/pages/includes/tctl.mdx!)
- A host where you will run the Teleport Application Service.

> - A DNS subdomain for each application.

Reviewer 2: I don't think there's enough information in this bullet. A subdomain of which domain? What does "each application" refer to? And what does it mean to have a subdomain?

>    <Admonition type="note" title="Application Access and DNS for Self-Hosted Clusters">

Reviewer 2: I would remove the `Admonition` tags and incorporate the text into the bullet above. Otherwise, this seems like special-case advice that readers can ignore, whereas, if a reader has Teleport Team/Enterprise Cloud, they will not need DNS subdomains.

   (!docs/pages/includes/dns-app-access.mdx!)

   Teleport Cloud automatically provides a subdomain and signed TLS certificate for 
   your application under your tenant address.
   </Admonition>

- Docker or Kubernetes to launch Grafana.

   <Tabs>
   <TabItem label="Docker">
   You must have Docker installed to launch any application in a Docker container. 
   If you are configuring access to any web application using Teleport, use the 

>    appropriate commands to launch that application in a Docker container.

Reviewer 2: Should the Docker container be launched before the user starts this guide? If so, why are we having the user run `docker run` with Grafana below?

    --app-uri=http://localhost:3000
```

## Step 2/5. Start Grafana

You can start Grafana in a Docker container or Kubernetes cluster by running the appropriate steps.

> For more information, click **Docker** or **Kubernetes cluster**.

Reviewer: ```suggestion
```

IMO we can trust the reader to know to click on the correct tab for their environment of choice.

<Tabs>
<TabItem label="Docker">

You can start Grafana in a Docker container by running the following command:

```code

> $ docker run -d -p 3000:3000 grafana/grafana

Reviewer: FYI The merge conflict that will need to be resolved is likely on this line, where we recently fixed this to provide a name for the container.

Author: Yep, I saw that go in.

```

After running this command, edit the `/etc/grafana/grafana.ini` file in the container. 
Under `[server]`, uncomment the `domain` key and set the value to your Teleport
cluster domain, (for example, `teleport.example.com`). 
Because the container doesn't have a text editor installed, you can use the `sed` command to edit the file. 

> For example:

Reviewer: As above, we line break around 80 chars and not for sentence ends.

```code
$ docker exec -u 0 grafana sed -i 's/;domain = localhost/domain = teleport.example.com/g' /etc/grafana/grafana.ini
```

Replace `teleport.example.com` with your Teleport cluster domain.
</TabItem>
<TabItem label="Kubernetes cluster">

If you install Grafana with the following `helm` instructions, it is available at `http://example-grafana.example-grafana.svc.cluster.local`
within the Kubernetes cluster.

```code
$ helm repo add grafana https://grafana.github.io/helm-charts
$ helm repo update
$ helm install example-grafana grafana/grafana \
    --create-namespace \
    --namespace example-grafana
```
</TabItem>
</Tabs>

## Step 3/5. Install and configure Teleport

Install Teleport on the host where you will run the Teleport Application Service.

<Tabs>
<TabItem label="Docker">
See the [Installation](../installation.mdx) page for options besides Linux servers.

> (!docs/pages/includes/install-linux.mdx!)

Reviewer: Unrelated to your edits, just something I need to get off my chest: I don't get why we don't provide instructions here to install the Teleport Application Service in another Docker container if we're already using Docker for Grafana.

Author: I'll bite: what are the instructions?

Reviewer: Something like:

```
docker run --name teleport \
  -v ~/teleport/config:/etc/teleport \
  public.ecr.aws/gravitational/teleport-distroless:13.0.4
```

Author: Prerequisites say you have a running cluster and the Docker instructions are covered in the Installation page, so I think I'll skip adding them here for now. I might revisit this page because the process is easier, less error-prone, and more generic through the web UI.

Create the Teleport Application Service configuration at `/etc/teleport.yaml`.
Replace <Var name="teleport.example.com" /> with your Teleport Proxy Service

> address or Teleport Cloud tenant (for example, `mytenant.teleport.sh`).

Reviewer: Here and below, why remove the example of a self-hosted proxy address?

Author: It seemed clutter-y to keep repeating `teleport.example.com` everywhere and I'm not sure what the rationale is for using different terms for the `teleport.example.com` address (Teleport cluster name, Teleport Proxy Service address, domain, URL).

```code
$ docker run --detach --name grafana --publish 3000:3000 grafana/grafana
```

After running this command, edit the `/etc/grafana/grafana.ini` file in the container. 
Under `[server]`, uncomment the `domain` key and set the value to your Teleport
cluster domain, for example, `teleport.example.com`. Because the container doesn't 
have a text editor installed, you can use the `sed` command to edit the file. 
For example:

```code
$ docker exec --user 0 grafana sed -i \
  's/;domain = localhost/domain = teleport.example.com/g' \
  /etc/grafana/grafana.ini
```

Replace `teleport.example.com` with your Teleport cluster domain.
</TabItem>
<TabItem label="Kubernetes cluster">

If you install Grafana with the following `helm` instructions, it is available 
at `http://example-grafana.example-grafana.svc.cluster.local` within the 
Kubernetes cluster.

```code
$ helm repo add grafana https://grafana.github.io/helm-charts
$ helm repo update
$ helm install example-grafana grafana/grafana \
    --create-namespace \
    --namespace example-grafana
```
</TabItem>
</Tabs>

## Step 3/5. Install and configure Teleport

> Install Teleport on the host that will run the Teleport Application Service.

Reviewer 2: Why are we including this paragraph when there's a `TabItem` below that shows you how to run the Teleport Application Service in a Kubernetes pod?

  --app-uri=http://localhost:3000
```
  
- Set the `--proxy` command-line option to your Teleport Proxy Service address 
(for example, `teleport.example.com`) or Teleport Cloud tenant 
(for example, `mytenant.teleport.sh`).
- Set the `--token` command-line option to the file location on the Application 

> Service host where you saved the invitation token that you generated earlier.

Reviewer 2: Would there be an Application Service host if you're running the service via the `teleport-kube-agent` chart? Would it make sense to make this value the value of the token instead?

- Change the `--app-name` and `--app-uri` command-line options if you're configuring 
access to a different web application.

(!docs/pages/includes/start-teleport.mdx service="the Teleport Application Service"!)
</TabItem>
<TabItem label="Kubernetes cluster">

(!docs/pages/kubernetes-access/helm/includes/helm-repo-add.mdx!)

> <ScopedBlock scope={["oss", "enterprise"]}>

Reviewer 2: What's the difference between this `ScopedBlock` and the one below it? Should we merge them into one?

(!docs/pages/kubernetes-access/helm/includes/helm-repo-add.mdx!)

> Install the `teleport` pod running Grafana into your Kubernetes cluster
> with a configuration similar to the following:

Reviewer 2: ```suggestion
Install the `teleport-kube-agent` Helm chart into your Kubernetes cluster to proxy Grafana
with a command similar to the following:
```
The `teleport` pod doesn't run Grafana.

  --version (=teleport.version=)
```

- Set `proxyAddr` to your Teleport Proxy Service address (for example, 
`teleport.example.com`) or Teleport Cloud tenant (for example, `mytenant.teleport.sh`).
- Set `-authToken` to the file location on the Application Service host where you 
saved the invitation token that you generated earlier.
- Change `apps[0].name` and `apps[0].uri` if you're configuring access to a different 
web application.

</ScopedBlock>
<ScopedBlock scope={["cloud"]}>

> Install the Teleport Kubernetes Service into your Kubernetes cluster

Reviewer 2: Same comment as above: these instructions don't run the Kubernetes Service
